<a href="https://colab.research.google.com/github/tortas/DS-Unit-1-Sprint-2-Data-Wrangling/blob/master/module1-scrape-and-process-data/LS_DS_121_Scrape_and_process_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

_Lambda School Data Science_

# Scrape and process data

Objectives
- scrape and parse web pages
- use list comprehensions
- select rows and columns with pandas

Links
-  [Automate the Boring Stuff with Python, Chapter 11](https://automatetheboringstuff.com/chapter11/)
  - Requests
  - Beautiful Soup
- [Python List Comprehensions: Explained Visually](https://treyhunner.com/2015/12/python-list-comprehensions-now-in-color/)
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
  - Subset Observations (Rows)
  - Subset Variables (Columns)
- Python Data Science Handbook
  - [Chapter 3.1](https://jakevdp.github.io/PythonDataScienceHandbook/03.01-introducing-pandas-objects.html), Introducing Pandas Objects
  - [Chapter 3.2](https://jakevdp.github.io/PythonDataScienceHandbook/03.02-data-indexing-and-selection.html), Data Indexing and Selection


## Scrape the titles of PyCon 2019 talks

In [0]:
url = 'https://us.pycon.org/2019/schedule/talks/list/'

## 5 ways to look at long titles

Let's define a long title as greater than 80 characters

### 1. For Loop

### 2. List Comprehension

### 3. Filter with named function

### 4. Filter with anonymous function

### 5. Pandas

pandas documentation: [Working with Text Data](https://pandas.pydata.org/pandas-docs/stable/text.html)

## Make new dataframe columns

pandas documentation: [apply](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.apply.html)

### title length

### long title

### first letter

### word count

Using [`textstat`](https://github.com/shivam5992/textstat)

In [0]:
!pip install textstat

## Rename column

`title length` --> `title character count`

pandas documentation: [rename](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename.html)

## Analyze the dataframe

### Describe

pandas documentation: [describe](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.describe.html)

### Sort values

pandas documentation: [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

Five shortest titles, by character count

Titles sorted reverse alphabetically

### Get value counts

pandas documentation: [value_counts](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)


Frequency counts of first letters

Percentage of talks with long titles

### Plot

pandas documentation: [Visualization](https://pandas.pydata.org/pandas-docs/stable/visualization.html)





Top 5 most frequent first letters

Histogram of title lengths, in characters

# Assignment

**Scrape** the talk descriptions. Hint: `soup.select('.presentation-description')`

**Make** new columns in the dataframe:
- description
- description character count
- description word count

**Describe** all the dataframe's columns. What's the average description word count? The minimum? The maximum?

**Answer** the question: Which descriptions could fit in a tweet?


# Stretch Challenge

**Make** another new column in the dataframe:
- description grade level (you can use [this `textstat` function](https://github.com/shivam5992/textstat#the-flesch-kincaid-grade-level) to get the Flesh-Kincaid grade level)

**Answer** the question: What's the distribution of grade levels? Plot a histogram.

**Be aware** that [Textstat has issues when sentences aren't separated by spaces](https://github.com/shivam5992/textstat/issues/77#issuecomment-453734048). (A Lambda School Data Science student helped identify this issue, and emailed with the developer.) 

Also, [BeautifulSoup doesn't separate paragraph tags with spaces](https://bugs.launchpad.net/beautifulsoup/+bug/1768330).

So, you may get some inaccurate or surprising grade level estimates here. Don't worry, that's ok — but optionally, can you do anything to try improving the grade level estimates?

In [0]:
#BEGIN ASSIGNMENT 121 CODE

In [0]:
import pandas as pd
url = 'https://us.pycon.org/2019/schedule/talks/list/'

In [0]:
import bs4
import requests

result = requests.get(url)

In [0]:
pd.options.display.max_colwidth = 200
soup = bs4.BeautifulSoup(result.text)
descriptions = [desc.text.strip()
                for desc in soup.select('.presentation-description')]

In [0]:
df = pd.DataFrame({'description': descriptions})

In [8]:
!pip install textstat
import textstat

    100% |████████████████████████████████| 3.0MB 7.8MB/s 


In [0]:
df['description character count'] = df['description'].apply(len)

In [0]:
df['description word count'] = df['description'].apply(textstat.lexicon_count)

In [12]:
df.describe(include='all')

description  \
count                                                                                                                                                                                                        95   
unique                                                                                                                                                                                                       95   
top     As teams and projects grow, code review becomes increasingly important to support the maintainability of complex code bases. In this talk, I’ll cover guidelines for writing consistent code, powerf...   
freq                                                                                                                                                                                                          1   
mean                                                                                                                                                                                                        NaN   
std                                                                                                                                                                                                         NaN   
min                                                                                                                                                                                                         NaN   
25%                                                                                                                                                                                                         NaN   
50%                                                                                                                                                                                                         NaN   
75%                                                                                                                                                                                                         NaN   
max                                                                                                                                                                                                         NaN   

        description character count  description word count  
count                     95.000000               95.000000  
unique                          NaN                     NaN  
top                             NaN                     NaN  
freq                            NaN                     NaN  
mean                     813.073684              130.821053  
std                      415.988191               64.357872  
min                      121.000000               20.000000  
25%                      542.500000               85.500000  
50%                      718.000000              116.000000  
75%                     1016.500000              165.000000  
max                     2827.000000              421.000000

*   Average description word count: About 131
*   Word count minimum: 20
*   Word count maximum: 421



In [15]:
df [df['description character count'] <= 280]

description  \
54  Learn how to make music with Python, SuperCollider and FoxDot. We'll create a song together in this live coded adventure.   

    description character count  description word count  
54                          121                      20

It looks like there's only one description that would fit under twitter's
280 character limit! 